% Template for PLoS
% Version 1.0 January 2009
%
% To compile to pdf, run:
% latex plos.template
% bibtex plos.template
% latex plos.template
% latex plos.template
% dvipdf plos.template

\documentclass[10pt]{article}

% amsmath package, useful for mathematical formulas
\usepackage{amsmath}
% amssymb package, useful for mathematical symbols
\usepackage{amssymb}

% graphicx package, useful for including eps and pdf graphics
% include graphics with the command \includegraphics
\usepackage{graphicx}

% cite package, to clean up citations in the main text. Do not remove.
\usepackage{cite}

\usepackage{color} 

% Use doublespacing - comment out for single spacing
%\usepackage{setspace} 
%\doublespacing


% Text layout
\topmargin 0.0cm
\oddsidemargin 0.5cm
\evensidemargin 0.5cm
\textwidth 16cm 
\textheight 21cm

% Bold the 'Figure #' in the caption and separate it with a period
% Captions will be left justified
\usepackage[labelfont=bf,labelsep=period,justification=raggedright]{caption}

% Use the PLoS provided bibtex style
\bibliographystyle{plos2009}

% Remove brackets from numbering in List of References
\makeatletter
\renewcommand{\@biblabel}[1]{\quad#1.}
\makeatother


% Leave date blank
\date{}

\pagestyle{myheadings}
%% ** EDIT HERE **


%% ** EDIT HERE **
%% PLEASE INCLUDE ALL MACROS BELOW

%% END MACROS SECTION

\begin{document}

% Title must be 150 characters or less
\begin{flushleft}
{\Large
\textbf{Metagenome phylogeny for the 1\%}
}
% Insert Author names, affiliations and corresponding author email.
\\
Person A$^{1}$, 
Person B$^{1,2,3}$, 
Person C$^{1,\ast}$
\\
\bf{1} Genome Center, University of California-Davis, California, United States of America
\\
\bf{2} Department of Evolution and Ecology, University of California-Davis, California, United States of America
\\
\bf{3} Department. of Medical Microbiology and Immunology, University of California-Davis, California, United States of America
\\
$\ast$ E-mail: Corresponding aarondarling@ucdavis.edu
\end{flushleft}

% Please keep the abstract between 250 and 300 words
\section*{Abstract}
blahdeeblah, blah blah
% Please keep the Author Summary between 150 and 200 words
% Use first person. PLoS ONE authors please skip this step. 
% Author Summary not valid for PLoS ONE submissions.   
\section*{Author Summary}

\section*{Introduction}

We follow in the footsteps of the great Martin Wu and Jonathan Eisen~\cite{WuEisen2008}.

\subsection*{Previous work}
Microbial ecology is a burgeoning field, we focus here on shotgun metagenome sequencing approaches and 
do not discuss analysis methods for amplicon sequencing data such as QIIME and others.
\subsubsection*{Whole metagenome binning}
- composition classifiers
  - TACOA?, PhyloPythia 1 \& 2
- identity/homology classifiers
  - MEGAN, SORT-Items, others
\subsubsection*{Estimating community composition from metagenomes}
      - AMPHORA, MLTreeMap, others?
      - MetaPHlan

% You may title this section "Methods" or "Models". 
% "Models" is not a valid title for PLoS ONE authors. However, PLoS ONE
% authors may use "Analysis" 
\section*{Methods or Design and Implementation}

\subsection*{Detailed PhyloSift workflow}
Stream computing
\subsubsection*{Sequence identity search}
Process sequence (reads or assembly) to identify similarity to marker genes
      - Evaluated BLAST, BLAST+, RAPsearch2, LAST, bowtie2. Picked LAST, bowtie2.
\subsubsection*{Alignment to reference multiple alignment}
      - reverse translation to DNA when possible
\subsubsection*{Placement on a phylogenetic reference tree}

\subsubsection*{Taxonomic summary of read placements}
      - taxonomic summary
      - other summaries
\subsubsection*{Comparison among samples}

\subsection*{PhyloSift database update workflow}
\subsubsection*{Acquire new genome data}
\subsubsection*{Run PhyloSift workflow on each genome}
\subsubsection*{Taxonomic reconciliation}

% Results and Discussion can be combined.
\section*{Results}


\subsection*{look Ma, I'm famous}

\subsection*{Simulated data}
    - How they were generated
    - Building PhyloSift DBs without test datasets
\subsubsection*{Precision, recall, and F1}
\subsubsection*{Relationship between F1 and neighbor taxon density}

\subsubsection*{Community structure accuracy}
RMSD or other metric?

\subsubsection*{Accuracy on particular taxonomic groups}
Bacteria, Arch, Eukarya, viral

\subsection*{Application to a real dataset}
 (which one?)

\subsection*{PhyloSift crushes its former self}


\section*{Discussion}
  - Fundamental limits to computational methods -- resolving linkage among polymorphisms in a population
  - Advantages to the approach taken by PhyloSift
  - Disadvantages
    - many


\subsection*{Limitations and scope}


\subsection*{Future work}


\section*{Availability}
Software for Linux and Mac OS X, along with source code is freely available from http://github.com/gjospin/PhyloSift
The source code has been licensed under the GNU Public License (GPL) v3.0.

% Do NOT remove this, even if you are not including acknowledgments
\section*{Acknowledgments}
Funding from Dept. of Homeland Security, DOE.

%\section*{References}
% The bibtex filename
\bibliography{phylosift}

\clearpage

\section*{Figure Legends}
%\begin{figure}[!ht]
%\begin{center}
%%\includegraphics[width=4in]{figure_name.2.eps}
%\end{center}
%\caption{
%{\bf Bold the first sentence.}  Rest of figure 2  caption.  Caption 
%should be left justified, as specified by the options to the caption 
%package.
%}
%\label{Figure_label}
%\end{figure}
\begin{figure}[hp]
\begin{center}
\end{center}
\caption{\textbf{A really pretty picture.} As you can see, the curves in this image are stunning.
Pick your jaw up off the floor, this ain't no joke.}
\label{fig:the_first_figure}
\end{figure}

\clearpage

\section*{Tables}


\clearpage

\end{document}

